src/common/bitbuffer.cpp

```diff
@@ -0,0 +1,185 @@
+    // exactly 0 component, number theoretically. (If it's not exactly 0 by number theory, it's numerically
+    // negligible, and we're safe.)
+
+    BitOp outBuffer(new complex[4], std::default_delete<complex[]>());
+
+    if (rightBuffer != NULL) {
+        GateBuffer* rightGate = dynamic_cast<GateBuffer*>(rightBuffer.get());
+        BitOp right = rightGate->matrix;
+
+        std::vector<std::future<void>> futures(4);
+
+        futures[0] = std::async(std::launch::async, [&]() {
+            outBuffer.get()[0] = (matrix.get()[0] * right.get()[0]) + (matrix.get()[1] * right.get()[2]);
+            if (norm(outBuffer.get()[0]) < min_norm) {
+                outBuffer.get()[0] = complex(ZERO_R1, ZERO_R1);
+            }
+        });
+        futures[1] = std::async(std::launch::async, [&]() {
+            outBuffer.get()[1] = (matrix.get()[0] * right.get()[1]) + (matrix.get()[1] * right.get()[3]);
+            if (norm(outBuffer.get()[1]) < min_norm) {
+                outBuffer.get()[1] = complex(ZERO_R1, ZERO_R1);
+            }
+        });
+        futures[2] = std::async(std::launch::async, [&]() {
+            outBuffer.get()[2] = (matrix.get()[2] * right.get()[0]) + (matrix.get()[3] * right.get()[2]);
+            if (norm(outBuffer.get()[2]) < min_norm) {
+                outBuffer.get()[2] = complex(ZERO_R1, ZERO_R1);
+            }
+        });
+        futures[3] = std::async(std::launch::async, [&]() {
+            outBuffer.get()[3] = (matrix.get()[2] * right.get()[1]) + (matrix.get()[3] * right.get()[3]);
+            if (norm(outBuffer.get()[3]) < min_norm) {
+                outBuffer.get()[3] = complex(ZERO_R1, ZERO_R1);
+            }
+        });
+
+        for (int i = 0; i < 4; i++) {
+            futures[i].get();
+        }
+    } else {
+        std::copy(matrix.get(), matrix.get() + 4, outBuffer.get());
+    }
+
+    return std::make_shared<GateBuffer>(this, outBuffer);
+}
+
+void GateBuffer::Apply(QInterfacePtr qReg, const bitLenInt& qubitIndex, std::vector<BitBufferPtr>* bitBuffers)
+{
+    if (controls.size() == 0) {
+        qReg->ApplySingleBit(matrix.get(), true, qubitIndex);
+    } else {
+        bitLenInt* ctrls = new bitLenInt[controls.size()];
+        std::copy(controls.begin(), controls.end(), ctrls);
+
+        if (anti) {
+            qReg->ApplyAntiControlledSingleBit(ctrls, controls.size(), qubitIndex, matrix.get());
+        } else {
+            qReg->ApplyControlledSingleBit(ctrls, controls.size(), qubitIndex, matrix.get());
+        }
+
+        delete[] ctrls;
+    }
+    (*bitBuffers)[qubitIndex] = NULL;
+}
+
+bool ArithmeticBuffer::Combinable(BitBufferPtr toCmp)
+{
+    if (toCmp == NULL) {
+        return true;
+    }
+
+    if (BitBuffer::Combinable(toCmp) == false) {
+        return false;
+    }
+
+    ArithmeticBuffer* toCmpArith = dynamic_cast<ArithmeticBuffer*>(toCmp.get());
```

Comment:
Comment: "BitBuffer:: Combinable requires either both or neither of the buyers to be arithmetic, which makes this cast safe".

Reply:
Done.